// twilio-sync/karma.conf.js
module.exports = function(config) {
  config.set({
    browserStack: {
      username: process.env.BROWSERSTACK_USERNAME,
      accessKey: process.env.BROWSERSTACK_ACCESS_KEY
    },

    logLevel: config.LOG_INFO,
    autoWatch: false,
    singleRun: true,

    frameworks: ['mocha', 'browserify'],
    files: ['test/unit/index.js'],
    preprocessors: { 'test/unit/index.js': [ 'browserify' ] },

    browserify: {
      debug: true,
      bundleDelay: 1000,
      transform: [[ 'babelify', {
        global: true,
        ignore: /\/node_modules\/(?!twilio-transport|twilio-notifications|twilsock\/)/
      }]],
    },

    reporters: ['mocha'],

    // define browsers
    customLaunchers: {
      bs_firefox_mac: {
        base: 'BrowserStack',
        browser: 'firefox',
        browser_version: '21.0',
        os: 'OS X',
        os_version: 'Mountain Lion'
      },
      bs_iphone5: {
        base: 'BrowserStack',
        device: 'iPhone 5',
        os: 'ios',
        os_version: '6.0'
      }
    },

    browsers: ['bs_firefox_mac', 'bs_iphone5']
  })
}
